### DISCUSSION TITLE
getSession always returns null on server

### DISCUSSION THREAD

**Discussion by bratsos** [nextauthjs/next-auth#1466] [Help] [answered]
Hey community!

I'm trying to get the user's session on server side in order to render the correct page (login or dashboard) from the server. I'm using this on `pages/index.tsx`

```javascript
export const getServerSideProps = async ({ req }: NextPageContext) => {
  const session = await getSession({ req })

  console.log('SSR Session', session)
  return {
    props: { // doesn't even matter }
  }
}
```

but `session` is always `null` when it comes from server.

I'm also getting the following error in the console

```javascript

[next-auth][error][client_fetch_error]
https://next-auth.js.org/errors#client_fetch_error session FetchError: request to http://localhost:3000/api/auth/session failed, reason: connect ECONNREFUSED ::1:3000
```

If I replay the same GET request on the browser it works fine (and I'm getting the correct session as a response). I also tried to add a dumb endpoint `pages/api/hello-world` which is just replies with a 200 status code, and call this by directly editing `node_modules/next-auth/dist/client/index.js`. I'm getting the same `connect ECONNREFUSED ::1:3000` error.

I have `NEXTAUTH_URL=http://localhost:3000` in my `.env.local` and I can see the URL is correct by the error message in the console, plus I can verify that it works on client-side.

Am I missing something super obvious here?

Edit: I'm using `next-auth: 3.11.0`

**Comment by balazsorban44** [comment]
could you please provide a bit more information about your configuration and how you use signin/signout?

**Reply by bratsos** [reply]
Thanks for your reply!

My `[...nextauth].ts` file contains the following simple provider configuration

```javascript
const options = {
  providers: [
    Providers.Google({
      clientId: googleClientId,
      clientSecret: googleClientSecret,
    }),
    Providers.Facebook({
      clientId: facebookClientId,
      clientSecret: facebookClientSecret
    }),
    Providers.Credentials({
      name: 'Credentials',
      credentials: {
        email: { label: "e-mail", type: "text", placeholder: "jsmith@gmail.com" },
        password: { label: "Password", type: "password" }
      },
      async authorize(credentials) {
        // dumb authorization logic with my database
      }
    })
  ],
  callbacks: {
    async redirect(_url: string, baseUrl: string) {
      return baseUrl
    }
  },
  session: {
    jwt: true,
    maxAge: 30 * 24 * 60 * 60, // 30 days
  },
  database: mongoUrl,
}

export default (req: NextApiRequest, res: NextApiResponse) => {
  return NextAuth(req, res, options)
}
```

Login is triggered by manually invoking `signIn('facebook')` and `signIn('google')` functions


```javascript
// pages/index.tsx
<Login
        onGooglePress={() => signIn(Providers.google)} 
        onFacebookPress={() => signIn(Providers.facebook)}
/>
```

`Login` is just a UI wrapper of my login screen that renders 2 buttons.

I've managed to login successfully (and maintain the session after a full page refresh).

Feel free to ask for more details if I missed anything here, and thanks for the quick reply!

**Comment by iaincollins** [comment, marked as answer]
Things that may cause this include:

1. **Locally running firewall software that is preventing the server side route form connecting back to localhost (e.g. Windows Defender).**
  In this case you should be able to address this in the firewall settings.
  
2. **Running inside a virtual machine / container where "localhost" is relative to the container.**
  There is no quick solution for this one yet (but one is in progress), but if it does apply to you there are some old posts about how work around this.
  
The only other edge cases for this sort of thing I can think of are weird local DNS resolution problems caused by different resolvers for localhost in different contexts, but if that is the case then probably more thing are broken (so seems unlikely).

If neither 1 or 2 apply to you could try `http://127.0.0.1:3000` instead of `http://localhost:3000` as your `NEXTAUTH_URL` and see if that works any better for you.

**Reply by bratsos** [reply]
Neither 1 or 2 apply in my situation, but using `http://127.0.0.1:3000` as my `NEXTAUTH_URL` indeed works!

I'm using [expo + nextjs](https://docs.expo.io/guides/using-nextjs/) but I assumed that the web app won't be affected by any weird DNS resolution on expo's side - I can imagine this not working inside an emulator.

I'll try to find exactly what's happening (suggestions are welcomed) but thanks for pointing this out!

I'll mark this as resolved and come back if I find exactly what's happening.

**Reply by iaincollins** [reply]
Thanks for the update! That's really interesting. I wonder what is going on there!

Appreciate the follow up - this is useful to know about as it might be the sort of thing we could build in a check for this condition at some point to help folks out who run into similar problems.

**Reply by Kipitup** [reply]
Hello @iaincollins  did you manage to know why this work ?

I've just started a discussion around this issue (https://github.com/nextauthjs/next-auth/discussions/4870) and this worked for me. However, I find it really weird that it's only in production. In local dev (next dev), localhost:3000 is working perfectly well.

Any input ?

**Reply by Kipitup** [reply]
@bratsos including you in this follow up :)

**Reply by 0xAskar** [reply]
any update on this? same problem persisting after changing to http://127.0.0.1:3000

**Reply by poiintt** [reply]
Using `NEXTAUTH_URL_INTERNAL` with `http://localhost:3000` worked for me

**Reply by shirotamaki** [reply]
Thanks for your advice, it is much appreciated.
I have solved my problem with the following settings !
```
NEXTAUTH_URL=http://localhost:3000
NEXTAUTH_URL_INTERNAL=http://127.0.0.1:3000
```

**Reply by abhaysharma3021** [reply]
Hi, I am using getSession in an api which returns currentUser session, its working perfectly after login, i'm able to fetch user data but when I'm trying call another api in the app and checking currentUser exists its returing [next-auth][error][CLIENT_FETCH_ERROR] . Can somebody help me with this?

**Reply by Bulletninja** [reply]
Same issue as you @abhaysharma3021, where you able to solve it?

**Reply by jvmhonorato** [reply]
create file: ' .eslintrc.json ' in the root of the application with the following contents:

```
{
  "env":{"browser": true, "node":true, "es6":true},
  "extends": ["eslint:recommended","next/core-web-vitals"]
}

**Reply by Siddhesh-Shinde** [reply]
```
NEXTAUTH_URL=http://0.0.0.0:3000
NEXTAUTH_URL_INTERNAL=http://127.0.0.1:3000
```
This worked for me as well

**Comment by nemanjam** [comment]
This solves it https://next-auth.js.org/configuration/nextjs#in-api-routes

```ts
import { authOptions } from 'pages/api/auth/[...nextauth]'
import { getServerSession } from "next-auth/next"


export async function handler(req, res) {
  const session = await getServerSession(req, res, authOptions)
...
```

**Reply by wowczarczyk** [reply]
@balazsorban44  Do you think we could have a separate definition of the `getServerSession` method that takes `NextRequest` and `NextResponse` from the new Next.js 13 Route handlers? The problem with getting cookies and headers from `require("next/headers")` is that it won't work for standalone Next.js apps (on Docker), it only works in dev mode and presumably on Vercel.

**Reply by wowczarczyk** [reply]
Something like this works for me:

```
const req = {
    headers: Object.fromEntries(request.headers),
    cookies: Object.fromEntries(
      request.cookies.getAll().map((c) => [c.name, c.value])
    ),
  };
  const res = { getHeader() {}, setCookie() {}, setHeader() {} };

  const session = await getServerSession(req as any, res as any, authOptions);
  ```
  
  But obviously is hacky due to the `as any` cast...

**Comment by zilrehman** [comment]
If you are not able to get session use gettoken instead
In authOption file 
```
 secret: process.env.NEXTAUTH_SECRET as string,
  callbacks: {
                   async signIn(user:any) {
                              if (user) {
                                const data = await axios.post(`${process.env.NEXTAUTH_URL}/api/sign_in_create_account`, user);
                                if (data) {
                                  return user;
                                }
                            } else {
                              // Return false if login failed
                              console.log("Login failed");
                              return "/auth/signup";
                            }
                  },
                  async jwt({ token, account, profile,user }) {
                    // Persist the OAuth access_token and or the user id to the token right after signin
                    if (account) {
                      token.accessToken = account.access_token
                       token.id = user.id
                    }
                  return token;
    },
```

Keys and URL
```
generate your NEXTAUTH_SECRET key from
# Only for production – generate one here: https://generate-secret.vercel.app/32
and Url for local
NEXTAUTH_URL=http://localhost:3000
```
this is api route for calling token
```import { NextRequest, NextResponse } from "next/server";
import { getToken } from "next-auth/jwt"
export const GET = async (req) => {
  
  const session = await getToken({ req, secret:process.env.NEXTAUTH_SECRET })

  return NextResponse.json(
      { message: "GET Property Error", session },
      { status: 200 }
    );
  }
  ```
and the in your layout/page
```
const MainLayout = ({ children }) => {
  // Data Calling
  const [session, setSession] = useState(false);

  useEffect(() => {
    const fetch = async () => {
      if (!session) {
        const response = await axios.get("/api/auth/token");
        if (response.data.session) {
          setSession(true);
        }
      }
    };
    fetch();
  });
return (
    <>
      {session ? (
     // write code when session exist
) : (
     // write code when session is not there
      )}
    </>
  );
};

export default MainLayout;
```

**Reply by kuhlaid** [reply]
`NEXTAUTH_SECRET` needs to be set in your environment variables (at least in the version 4.24.8). For the longest time I was using `SECRET` as the environment variable name without issue, but somewhere along the line `NEXTAUTH_SECRET` became a requirement it seems (at least in the way I use it).

**Comment by Bakhtiar85** [comment]
Hi, everyone. 

I thought, it was the issue of Prisma but it is an issue with getSession. 
 I created [issue](https://github.com/prisma/prisma-examples/issues/5574#issue-2032214741)
Here, someone made extra effort.

**Comment by huypham50** [comment]
For anyone seeing this issue with Next.js, you should use `getServerSession` instead of `getSession`

**Comment by jvmhonorato** [comment]
create file: ' .eslintrc.json ' in the root of the application with the following contents:

```
{
  "env":{"browser": true, "node":true, "es6":true},
  "extends": ["eslint:recommended","next/core-web-vitals"]
}

**Comment by mcmiro** [comment]
Using `getServerAuthSession()` instead of getServerSession() helped me. The import is: `import { getServerAuthSession } from "@/server/auth";`
